### PR TITLE
Fix retry logic

### PR DESCRIPTION
**Pull request by kevmoo**
Closes https://github.com/dart-lang/pub/issues/1556

### CODE REVIEW

**Comment by kevmoo** [comment]
There was a bit of an oversight in https://github.com/dart-lang/pub/commit/932e76efcd813be84045da31d1c4e17b18b227f3

The innermost client – _PubHttpClient – **throws** on bad status codes. These exceptions blow right past the retry logic.

My solution – move the check & throw logic into a _ThrowingClient that sits outside of of the retry logic.

**Comment by kevmoo** [comment]
...at least one test would be a good idea...

**Comment by kevmoo** [comment]
Feel free to squash/merge and delete the branch if approve...

File: lib/src/http.dart [hunk 1]
```diff
@@ -216,10 +184,54 @@ final _pubClient = new _PubHttpClient();
 /// we're waiting for them to come back up.
 final _retriedHosts = new Set<String>();
 
+class _ThrowingClient extends http.BaseClient {
```
Comment (nex3) [inline lib/src/http.dart]: Please document this.

Comment (kevmoo) [reply lib/src/http.dart]: PTAL

**Comment by nex3** [comment]
I'll merge once Travis comes back successful.

**Comment by kevmoo** [comment]
Perfect!

**Pull request merged**